// firebase-tools/lib/mcp/tools/remoteconfig/rollback_template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rollback_template = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const rollback_1 = require("../../../remoteconfig/rollback");
exports.rollback_template = (0, tool_1.tool)({
    name: "rollback_template",
    description: "Rollback to a specific version of Remote Config template for a project",
    inputSchema: zod_1.z.object({
        version_number: zod_1.z
            .number()
            .describe("The version number to roll back to. This field is required."),
    }),
    annotations: {
        title: "Rollback remote config template",
        readOnlyHint: false,
    },
    _meta: {
        requiresAuth: true,
        requiresProject: true,
    },
}, async ({ version_number }, { projectId }) => {
    if (version_number === undefined) {
        return (0, util_1.mcpError)(`No version number specified in the rollback requests`);
    }
    return (0, util_1.toContent)(await (0, rollback_1.rollbackTemplate)(projectId, version_number));
});
