// @angular/compiler-cli/bundles/private/migrations.js

      import {createRequire as __cjsCompatRequire} from 'module';
      const require = __cjsCompatRequire(import.meta.url);
    
import {
  DiagnosticCategoryLabel,
  NgCompiler
} from "../chunk-I3BPPTZC.js";
import {
  CompilationMode,
  DtsMetadataReader,
  DynamicValue,
  FatalDiagnosticError,
  ImportManager,
  PartialEvaluator,
  PotentialImportKind,
  PotentialImportMode,
  Reference,
  ReferenceEmitKind,
  ReferenceEmitter,
  StaticInterpreter,
  SymbolKind,
  TypeScriptReflectionHost,
  createForwardRefResolver,
  extractDecoratorQueryMetadata,
  extractTemplate,
  findAngularDecorator,
  getAngularDecorators,
  getRootDirs,
  isShim,
  parseDecoratorInputTransformFunction,
  queryDecoratorNames,
  reflectObjectLiteral,
  unwrapExpression
} from "../chunk-IJMZQ3RN.js";
import "../chunk-LS5RJ5CS.js";
import {
  getFileSystem,
  isLocalRelativePath
} from "../chunk-JEXAXD23.js";
import {
  NodeJSFileSystem
} from "../chunk-XYYEESKY.js";
import "../chunk-G7GFT6BU.js";
export {
  CompilationMode,
  DiagnosticCategoryLabel,
  DtsMetadataReader,
  DynamicValue,
  FatalDiagnosticError,
  ImportManager,
  NgCompiler,
  NodeJSFileSystem,
  PartialEvaluator,
  PotentialImportKind,
  PotentialImportMode,
  Reference,
  ReferenceEmitKind,
  ReferenceEmitter,
  StaticInterpreter,
  SymbolKind,
  TypeScriptReflectionHost,
  createForwardRefResolver,
  extractDecoratorQueryMetadata,
  extractTemplate,
  findAngularDecorator,
  getAngularDecorators,
  getFileSystem,
  getRootDirs,
  isLocalRelativePath,
  isShim,
  parseDecoratorInputTransformFunction,
  queryDecoratorNames,
  reflectObjectLiteral,
  unwrapExpression
};
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.dev/license
 */
