// @angular/build/src/builders/unit-test/schema.js
"use strict";
// THIS FILE IS AUTOMATICALLY GENERATED. TO UPDATE THIS FILE YOU NEED TO CHANGE THE
// CORRESPONDING JSON SCHEMA FILE, THEN RUN devkit-admin build (or bazel build ...).
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runner = exports.CoverageReporterEnum = void 0;
var CoverageReporterEnum;
(function (CoverageReporterEnum) {
    CoverageReporterEnum["Cobertura"] = "cobertura";
    CoverageReporterEnum["Html"] = "html";
    CoverageReporterEnum["Json"] = "json";
    CoverageReporterEnum["JsonSummary"] = "json-summary";
    CoverageReporterEnum["Lcov"] = "lcov";
    CoverageReporterEnum["Lcovonly"] = "lcovonly";
    CoverageReporterEnum["Text"] = "text";
    CoverageReporterEnum["TextSummary"] = "text-summary";
})(CoverageReporterEnum || (exports.CoverageReporterEnum = CoverageReporterEnum = {}));
/**
 * Specifies the test runner to use for test execution.
 */
var Runner;
(function (Runner) {
    Runner["Karma"] = "karma";
    Runner["Vitest"] = "vitest";
})(Runner || (exports.Runner = Runner = {}));
