// firebase-tools/lib/mcp/tools/crashlytics/add_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.add_note = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const addNote_1 = require("../../../crashlytics/addNote");
const constants_1 = require("./constants");
exports.add_note = (0, tool_1.tool)({
    name: "add_note",
    description: "Add a note to an issue from crashlytics.",
    inputSchema: zod_1.z.object({
        app_id: constants_1.APP_ID_FIELD,
        issue_id: zod_1.z.string().describe("The issue id to add the note to."),
        note: zod_1.z.string().describe("The note to add to the issue."),
    }),
    annotations: {
        title: "Add note to Crashlytics issue.",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
    },
}, async ({ app_id, issue_id, note }) => {
    if (!app_id)
        return (0, util_1.mcpError)(`Must specify 'app_id' parameter.`);
    if (!issue_id)
        return (0, util_1.mcpError)(`Must specify 'issue_id' parameter.`);
    if (!note)
        return (0, util_1.mcpError)(`Must specify 'note' parameter.`);
    return (0, util_1.toContent)(await (0, addNote_1.addNote)(app_id, issue_id, note));
});
