// firebase-tools/lib/mcp/tools/crashlytics/delete_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.delete_note = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const deleteNote_1 = require("../../../crashlytics/deleteNote");
const constants_1 = require("./constants");
exports.delete_note = (0, tool_1.tool)({
    name: "delete_note",
    description: "Delete a note from an issue in Crashlytics.",
    inputSchema: zod_1.z.object({
        app_id: constants_1.APP_ID_FIELD,
        issue_id: zod_1.z.string().describe("The issue id to delete the note from."),
        note_id: zod_1.z.string().describe("The note id to delete."),
    }),
    annotations: {
        title: "Delete Note from Crashlytics Issue.",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
    },
}, async ({ app_id, issue_id, note_id }) => {
    if (!app_id) {
        return (0, util_1.mcpError)(`Must specify 'app_id' parameter.`);
    }
    if (!issue_id) {
        return (0, util_1.mcpError)(`Must specify 'issue_id' parameter.`);
    }
    if (!note_id) {
        return (0, util_1.mcpError)(`Must specify 'note_id' parameter.`);
    }
    return (0, util_1.toContent)(await (0, deleteNote_1.deleteNote)(app_id, issue_id, note_id));
});
