// firebase-tools/lib/mcp/tools/crashlytics/list_notes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list_notes = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const listNotes_1 = require("../../../crashlytics/listNotes");
const constants_1 = require("./constants");
exports.list_notes = (0, tool_1.tool)({
    name: "list_notes",
    description: "List all notes for an issue in Crashlytics.",
    inputSchema: zod_1.z.object({
        app_id: constants_1.APP_ID_FIELD,
        issue_id: zod_1.z.string().describe("The issue id to list notes for."),
        note_count: zod_1.z
            .number()
            .optional()
            .default(10)
            .describe("Number of notes that needs to be fetched. Defaults to 10 if unspecified."),
    }),
    annotations: {
        title: "List notes for a Crashlytics issue.",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
    },
}, async ({ app_id, issue_id, note_count }) => {
    if (!app_id) {
        return (0, util_1.mcpError)(`Must specify 'app_id' parameter.`);
    }
    if (!issue_id) {
        return (0, util_1.mcpError)(`Must specify 'issue_id' parameter.`);
    }
    note_count !== null && note_count !== void 0 ? note_count : (note_count = 10);
    return (0, util_1.toContent)(await (0, listNotes_1.listNotes)(app_id, issue_id, note_count));
});
