// firebase-tools/lib/mcp/tools/crashlytics/list_top_issues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list_top_issues = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const listTopIssues_1 = require("../../../crashlytics/listTopIssues");
const constants_1 = require("./constants");
exports.list_top_issues = (0, tool_1.tool)({
    name: "list_top_issues",
    description: "List the top crashes from crashlytics happening in the application.",
    inputSchema: zod_1.z.object({
        app_id: constants_1.APP_ID_FIELD,
        issue_count: zod_1.z
            .number()
            .optional()
            .describe("Number of issues that needs to be fetched. Defaults to 10 if unspecified.")
            .default(10),
        issue_type: zod_1.z
            .enum(["FATAL", "NON-FATAL", "ANR"])
            .optional()
            .default("FATAL")
            .describe("Types of issues that can be fetched comma-separated. Defaults to `FATAL` (Crashes). Other values include NON-FATAL (Non-fatal issues), ANR (Application not responding)."),
    }),
    annotations: {
        title: "List Top Crashlytics Issues.",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
    },
}, async ({ app_id, issue_type, issue_count }) => {
    if (!app_id)
        return (0, util_1.mcpError)(`Must specify 'app_id' parameter.`);
    issue_type !== null && issue_type !== void 0 ? issue_type : (issue_type = "FATAL");
    issue_count !== null && issue_count !== void 0 ? issue_count : (issue_count = 10);
    return (0, util_1.toContent)(await (0, listTopIssues_1.listTopIssues)(app_id, issue_type, issue_count));
});
