// firebase-tools/lib/mcp/tools/crashlytics/list_top_devices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list_top_devices = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const listTopDevices_1 = require("../../../crashlytics/listTopDevices");
const constants_1 = require("./constants");
exports.list_top_devices = (0, tool_1.tool)({
    name: "list_top_devices",
    description: "List the top devices from Crashlytics for an application.",
    inputSchema: zod_1.z.object({
        app_id: constants_1.APP_ID_FIELD,
        issue_id: zod_1.z.string().optional().describe("The issue id to filter on"),
        device_count: zod_1.z
            .number()
            .optional()
            .default(10)
            .describe("Number of devices that needs to be fetched. Defaults to 10 if unspecified."),
    }),
    annotations: {
        title: "List Top Crashlytics Devices.",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
    },
}, async ({ app_id, issue_id, device_count }) => {
    if (!app_id)
        return (0, util_1.mcpError)(`Must specify 'app_id' parameter.`);
    device_count !== null && device_count !== void 0 ? device_count : (device_count = 10);
    return (0, util_1.toContent)(await (0, listTopDevices_1.listTopDevices)(app_id, device_count, issue_id));
});
