// firebase-tools/lib/mcp/tools/crashlytics/update_issue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.update_issue = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const updateIssue_1 = require("../../../crashlytics/updateIssue");
const constants_1 = require("./constants");
exports.update_issue = (0, tool_1.tool)({
    name: "update_issue",
    description: "Update the state of an issue in Crashlytics.",
    inputSchema: zod_1.z.object({
        app_id: constants_1.APP_ID_FIELD,
        issue_id: zod_1.z.string().describe("The issue id to update."),
        state: zod_1.z
            .nativeEnum(updateIssue_1.IssueState)
            .describe("The new state for the issue. Can be 'OPEN' or 'CLOSED'."),
    }),
    annotations: {
        title: "Update Crashlytics issue state.",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
    },
}, async ({ app_id, issue_id, state }) => {
    if (!app_id) {
        return (0, util_1.mcpError)(`Must specify 'app_id' parameter.`);
    }
    if (!issue_id) {
        return (0, util_1.mcpError)(`Must specify 'issue_id' parameter.`);
    }
    if (!state) {
        return (0, util_1.mcpError)(`Must specify 'state' parameter.`);
    }
    return (0, util_1.toContent)(await (0, updateIssue_1.updateIssue)(app_id, issue_id, state));
});
