// @schematics/angular/ng-new/schema.js
"use strict";
// THIS FILE IS AUTOMATICALLY GENERATED. TO UPDATE THIS FILE YOU NEED TO CHANGE THE
// CORRESPONDING JSON SCHEMA FILE, THEN RUN devkit-admin build (or bazel build ...).
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewEncapsulation = exports.TestRunner = exports.Style = exports.PackageManager = exports.FileNameStyleGuide = exports.AiConfig = void 0;
var AiConfig;
(function (AiConfig) {
    AiConfig["Agents"] = "agents";
    AiConfig["Claude"] = "claude";
    AiConfig["Copilot"] = "copilot";
    AiConfig["Cursor"] = "cursor";
    AiConfig["Gemini"] = "gemini";
    AiConfig["Jetbrains"] = "jetbrains";
    AiConfig["None"] = "none";
    AiConfig["Windsurf"] = "windsurf";
})(AiConfig || (exports.AiConfig = AiConfig = {}));
/**
 * The file naming convention to use for generated files. The '2025' style guide (default)
 * uses a concise format (e.g., `app.ts` for the root component), while the '2016' style
 * guide includes the type in the file name (e.g., `app.component.ts`). For more
 * information, see the Angular Style Guide (https://angular.dev/style-guide).
 */
var FileNameStyleGuide;
(function (FileNameStyleGuide) {
    FileNameStyleGuide["The2016"] = "2016";
    FileNameStyleGuide["The2025"] = "2025";
})(FileNameStyleGuide || (exports.FileNameStyleGuide = FileNameStyleGuide = {}));
/**
 * The package manager used to install dependencies.
 */
var PackageManager;
(function (PackageManager) {
    PackageManager["Bun"] = "bun";
    PackageManager["Npm"] = "npm";
    PackageManager["Pnpm"] = "pnpm";
    PackageManager["Yarn"] = "yarn";
})(PackageManager || (exports.PackageManager = PackageManager = {}));
/**
 * The type of stylesheet files to be created for components in the initial project.
 */
var Style;
(function (Style) {
    Style["Css"] = "css";
    Style["Less"] = "less";
    Style["Sass"] = "sass";
    Style["Scss"] = "scss";
    Style["Tailwind"] = "tailwind";
})(Style || (exports.Style = Style = {}));
/**
 * The unit testing runner to use.
 */
var TestRunner;
(function (TestRunner) {
    TestRunner["Karma"] = "karma";
    TestRunner["Vitest"] = "vitest";
})(TestRunner || (exports.TestRunner = TestRunner = {}));
/**
 * Sets the view encapsulation mode for components in the initial project. This determines
 * how component styles are scoped and applied. Options include: `Emulated` (default, styles
 * are scoped to the component), `None` (styles are global), and `ShadowDom` (styles are
 * encapsulated using Shadow DOM).
 */
var ViewEncapsulation;
(function (ViewEncapsulation) {
    ViewEncapsulation["Emulated"] = "Emulated";
    ViewEncapsulation["None"] = "None";
    ViewEncapsulation["ShadowDom"] = "ShadowDom";
})(ViewEncapsulation || (exports.ViewEncapsulation = ViewEncapsulation = {}));
