// firebase-tools/lib/mcp/tools/auth/disable_user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disable_user = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const auth_1 = require("../../../gcp/auth");
exports.disable_user = (0, tool_1.tool)({
    name: "disable_user",
    description: "Disables or enables a user based on a UID.",
    inputSchema: zod_1.z.object({
        uid: zod_1.z.string().describe("The localId or UID of the user to disable or enable"),
        disabled: zod_1.z.boolean().describe("true disables the user, false enables the user"),
    }),
    annotations: {
        title: "Disable or enable a particular user",
        destructiveHint: true,
        idempotentHint: true,
    },
    _meta: {
        requiresAuth: true,
        requiresProject: true,
    },
}, async ({ uid, disabled }, { projectId }) => {
    const res = await (0, auth_1.disableUser)(projectId, uid, disabled);
    if (res) {
        return (0, util_1.toContent)(`User ${uid} as been ${disabled ? "disabled" : "enabled"}`);
    }
    return (0, util_1.toContent)(`Failed to ${disabled ? "disable" : "enable"} user ${uid}`);
});
