// firebase-tools/lib/mcp/tools/storage/get_rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get_rules = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const rules_1 = require("../../../gcp/rules");
exports.get_rules = (0, tool_1.tool)({
    name: "get_rules",
    description: "Retrieves the Firebase Cloud Storage Rules for the default bucket.",
    inputSchema: zod_1.z.object({}),
    annotations: {
        title: "Get Current Firebase Cloud Storage Rules",
        readOnlyHint: true,
    },
    _meta: {
        requiresProject: true,
        requiresAuth: true,
    },
}, async (_, { projectId }) => {
    const rulesetName = await (0, rules_1.getLatestRulesetName)(projectId, "firebase.storage");
    if (!rulesetName)
        return (0, util_1.mcpError)(`No active Firebase Storage rules were found in project '${projectId}'`);
    const rules = await (0, rules_1.getRulesetContent)(rulesetName);
    return (0, util_1.toContent)(rules[0].content);
});
