// @schematics/angular/application/schema.js
"use strict";
// THIS FILE IS AUTOMATICALLY GENERATED. TO UPDATE THIS FILE YOU NEED TO CHANGE THE
// CORRESPONDING JSON SCHEMA FILE, THEN RUN devkit-admin build (or bazel build ...).
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewEncapsulation = exports.TestRunner = exports.Style = exports.FileNameStyleGuide = void 0;
/**
 * The file naming convention to use for generated files. The '2025' style guide (default)
 * uses a concise format (e.g., `app.ts` for the root component), while the '2016' style
 * guide includes the type in the file name (e.g., `app.component.ts`). For more
 * information, see the Angular Style Guide (https://angular.dev/style-guide).
 */
var FileNameStyleGuide;
(function (FileNameStyleGuide) {
    FileNameStyleGuide["The2016"] = "2016";
    FileNameStyleGuide["The2025"] = "2025";
})(FileNameStyleGuide || (exports.FileNameStyleGuide = FileNameStyleGuide = {}));
/**
 * The type of stylesheet files to be created for components in the application.
 */
var Style;
(function (Style) {
    Style["Css"] = "css";
    Style["Less"] = "less";
    Style["Sass"] = "sass";
    Style["Scss"] = "scss";
    Style["Tailwind"] = "tailwind";
})(Style || (exports.Style = Style = {}));
/**
 * The unit testing runner to use.
 */
var TestRunner;
(function (TestRunner) {
    TestRunner["Karma"] = "karma";
    TestRunner["Vitest"] = "vitest";
})(TestRunner || (exports.TestRunner = TestRunner = {}));
/**
 * Sets the view encapsulation mode for the application's components. This determines how
 * component styles are scoped and applied.
 */
var ViewEncapsulation;
(function (ViewEncapsulation) {
    ViewEncapsulation["Emulated"] = "Emulated";
    ViewEncapsulation["None"] = "None";
    ViewEncapsulation["ShadowDom"] = "ShadowDom";
})(ViewEncapsulation || (exports.ViewEncapsulation = ViewEncapsulation = {}));
