// firebase-tools/lib/crashlytics/addNote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addNote = void 0;
const logger_1 = require("../logger");
const error_1 = require("../error");
const utils_1 = require("./utils");
async function addNote(appId, issueId, note) {
    const requestProjectNumber = (0, utils_1.parseProjectNumber)(appId);
    logger_1.logger.debug(`[mcp][crashlytics] addNote called with appId: ${appId}, issueId: ${issueId}, note: ${note}`);
    try {
        const response = await utils_1.CRASHLYTICS_API_CLIENT.request({
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            path: `/projects/${requestProjectNumber}/apps/${appId}/issues/${issueId}/notes`,
            body: { body: note },
            timeout: utils_1.TIMEOUT,
        });
        return response.body;
    }
    catch (err) {
        logger_1.logger.debug(err.message);
        throw new error_1.FirebaseError(`Failed to add note to issue ${issueId} for app ${appId}. Error: ${err}.`, { original: err });
    }
}
exports.addNote = addNote;
