// firebase-tools/lib/crashlytics/getSampleCrash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSampleCrash = void 0;
const logger_1 = require("../logger");
const error_1 = require("../error");
const utils_1 = require("./utils");
async function getSampleCrash(appId, issueId, sampleCount, variantId) {
    const requestProjectNumber = (0, utils_1.parseProjectNumber)(appId);
    logger_1.logger.debug(`[mcp][crashlytics] getSampleCrash called with appId: ${appId}, issueId: ${issueId}, sampleCount: ${sampleCount}, variantId: ${variantId}`);
    try {
        const queryParams = new URLSearchParams();
        queryParams.set("filter.issue.id", issueId);
        queryParams.set("page_size", String(sampleCount));
        if (variantId) {
            queryParams.set("filter.issue.variant_id", variantId);
        }
        logger_1.logger.debug(`[mcp][crashlytics] getSampleCrash query paramaters: ${queryParams}`);
        const response = await utils_1.CRASHLYTICS_API_CLIENT.request({
            method: "GET",
            headers: {
                "Content-Type": "application/json",
            },
            path: `/projects/${requestProjectNumber}/apps/${appId}/events`,
            queryParams: queryParams,
            timeout: utils_1.TIMEOUT,
        });
        return response.body;
    }
    catch (err) {
        logger_1.logger.debug(err.message);
        throw new error_1.FirebaseError(`Failed to fetch the same crash for the Firebase AppId ${appId}, IssueId ${issueId}. Error: ${err}.`, { original: err });
    }
}
exports.getSampleCrash = getSampleCrash;
