// firebase-tools/lib/crashlytics/listTopIssues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listTopIssues = void 0;
const logger_1 = require("../logger");
const error_1 = require("../error");
const utils_1 = require("./utils");
async function listTopIssues(appId, issueType, issueCount) {
    const requestProjectNumber = (0, utils_1.parseProjectNumber)(appId);
    try {
        const queryParams = new URLSearchParams();
        queryParams.set("page_size", `${issueCount}`);
        queryParams.set("filter.issue.error_types", `${issueType}`);
        logger_1.logger.debug(`[mcp][crashlytics] listTopIssues called with appId: ${appId}, issueType: ${issueType}, issueCount: ${issueCount}`);
        const response = await utils_1.CRASHLYTICS_API_CLIENT.request({
            method: "GET",
            headers: {
                "Content-Type": "application/json",
            },
            path: `/projects/${requestProjectNumber}/apps/${appId}/reports/topIssues`,
            queryParams: queryParams,
            timeout: utils_1.TIMEOUT,
        });
        return response.body;
    }
    catch (err) {
        logger_1.logger.debug(err.message);
        throw new error_1.FirebaseError(`Failed to fetch the top issues for the Firebase app id: ${appId}. Error: ${err}.`, { original: err });
    }
}
exports.listTopIssues = listTopIssues;
