// firebase-tools/lib/mcp/tools/rules/get_rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRulesTool = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const rules_1 = require("../../../gcp/rules");
function getRulesTool(productName, releaseName) {
    return (0, tool_1.tool)({
        name: "get_rules",
        description: `Retrieves the active ${productName} security rules for the current project.`,
        inputSchema: zod_1.z.object({}),
        annotations: {
            title: `Get Current ${productName} Rules`,
            readOnlyHint: true,
        },
        _meta: {
            requiresProject: true,
            requiresAuth: true,
        },
    }, async (_, { projectId }) => {
        const rulesetName = await (0, rules_1.getLatestRulesetName)(projectId, releaseName);
        if (!rulesetName)
            return (0, util_1.mcpError)(`No active ${productName} rules were found in project '${projectId}'`);
        const rules = await (0, rules_1.getRulesetContent)(rulesetName);
        return (0, util_1.toContent)(rules[0].content);
    });
}
exports.getRulesTool = getRulesTool;
